packages/block-editor/src/components/provider/with-registry-provider.js

```diff
@@ -0,0 +1,44 @@
+/**
+ * WordPress dependencies
+ */
+import { useState, useEffect } from '@wordpress/element';
+import { withRegistry, createRegistry, RegistryProvider } from '@wordpress/data';
+import { createHigherOrderComponent } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import { storeConfig } from '../../store';
+import applyMiddlewares from '../../store/middlewares';
+
+const withRegistryProvider = createHigherOrderComponent( ( WrappedComponent ) => {
+	return withRegistry( ( { useSubRegistry = true, registry, ...props } ) => {
+		// Disable reason, this rule conflicts with the React hooks rule (no conditionals)
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const [ subRegistry, updateRegistry ] = useState( null );
+
+		if ( ! useSubRegistry ) {
+			return <WrappedComponent registry={ registry } { ...props } />;
+		}
+
+		useEffect( () => {
+			const newRegistry = createRegistry( {}, registry );
```

Comment:
It took me a while to understand what's happening here:
- so it creates a new registry and points the passed registry as a parent
- overrides `core/block-editor`
- does some magic with middlewares